### PR TITLE
Move subtype caches to a separate global class

### PR DESCRIPTION
**Pull request by ilevkivskyi**
This refactoring doesn't have much value in itself, but will be helpful to simplify and speed-up protocols in fine grained mode, see #4790.

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
I am intentionally using class variables and class methods instead of a singleton because I think creating a singleton somewhere in `build.py` will only complicate the code.

**Comment by JukkaL** [comment]
My guess is that the travis failures are due to high memory use. You can try lowering travis parallelism. If it helps, then that's likely the cause.

**Comment by ilevkivskyi** [comment]
Yes, it was memory problem, thanks! But instead of reducing parallelism, I reset caches completely on every `build.build` (i.e. on everything that supposed to be a cold run, but apparently tests don't restart mypy completely).

File: mypy/build.py [hunk 1]
```diff
@@ -265,6 +266,7 @@ def _build(sources: List[BuildSource],
                            fscache=fscache)
 
     try:
+        TypeState.reset_all()
```
Comment (JukkaL) [inline mypy/build.py]: If we aren't running a fine-grained incremental build, I think that we can reset the state after the build (before returning the build results), since it won't be useful any more at than point but it will consume some memory. We should also reset the cache when shutting down the daemon, maybe in `reset()`. It's good manners to not leave global state lying around at the end of a program.

Style nit: Move this to be outside the try statement, since we don't expect this to fail.

File: mypy/nodes.py [hunk 2]
```diff
@@ -2062,8 +2058,7 @@ def __init__(self, names: 'SymbolTable', defn: ClassDef, module_name: str) -> No
         self.assuming = []
         self.assuming_proper = []
         self.inferring = []
-        self._cache = set()
-        self._cache_proper = set()
+        TypeState.add_caches(self)
```
Comment (JukkaL) [inline mypy/nodes.py]: What if this call would be optional? The TypeState API would be a bit simpler if `TypeState` would deal with adding entries as needed. Generally I'd prefer to keep public APIs as simple as possible and hide all logic within the implementation. Now the API is leaking an implementation detail (cache entry needs to exist for a class before it can be used for caches).

File: mypy/nodes.py [hunk 3]
```diff
@@ -2099,25 +2094,9 @@ def get_containing_type_info(self, name: str) -> 'Optional[TypeInfo]':
                 return cls
         return None
 
-    def record_subtype_cache_entry(self, left: 'mypy.types.Instance',
```
Comment (JukkaL) [inline mypy/nodes.py]: Good to have this logic moved away from this module!

File: mypy/nodes.py [hunk 4]
```diff
@@ -2099,25 +2094,9 @@ def get_containing_type_info(self, name: str) -> 'Optional[TypeInfo]':
...
-                                proper_subtype: bool = False) -> bool:
-        if not proper_subtype:
-            return (left, right) in self._cache
-        return (left, right) in self._cache_proper
-
     def reset_subtype_cache(self) -> None:
         for item in self.mro:
-            item._cache = set()
-            item._cache_proper = set()
+            TypeState.reset_caches(item)
```
Comment (JukkaL) [inline mypy/nodes.py]: We could simplify module dependencies a bit by moving this method to `TypeState`. This would have the benefit of `TypeInfo` not needing any knowledge of `TypeState` (together with the above idea). Conceptually then `TypeState` would only hold information about speeding up certain `TypeInfo` operations, but it would be separate from `TypeInfo`. It's better if TypeInfo doesn't need to be aware of operations available for TypeInfos.

In the future we can move additional 'transient' global state from TypeInfos to TypeState (or some other place), such as `inferring` and `assuming` (it's best to not do it in this PR to avoid making this big).

File: mypy/typestate.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+"""
+A shared state for all TypeInfo's that holds global cache and dependency information,
```
Comment (JukkaL) [inline mypy/typestate.py]: Can you make it a bit more explicit that this is mutable global state (basically global variables)?

Grammar nit: TypeInfos, not TypeInfo's.

File: mypy/typestate.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+from typing import Dict, Set, Tuple
+
+MYPY = False
+if MYPY:
+    from typing import ClassVar
+    from mypy.nodes import TypeInfo
+    from mypy.types import Instance
+
+
+class TypeState:
```
Comment (JukkaL) [inline mypy/typestate.py]: Add docstring. In the docstring, maybe discuss at least these things:

* Purpose of this class.
* How to avoid leaking global state: this needs to be explicitly cleared after a build has finished and after a daemon shutdown.
* Invariants: Resetting subtype caches for a class has no semantic effect (only perf).

File: mypy/typestate.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    # 'caches' and 'caches_proper' are subtype caches, implemented as sets of pairs
+    # of (subtype, supertype), where supertypes are instances of given TypeInfo.
+    # We need the caches, since subtype checks for structural types are very slow.
```
Comment (JukkaL) [inline mypy/typestate.py]: Explicitly mention that `caches_proper` if for proper subtype checks and `caches` for normal subtype checks.

File: mypy/typestate.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    caches = {}  # type: ClassVar[Dict[TypeInfo, Set[Tuple[Instance, Instance]]]]
```
Comment (JukkaL) [inline mypy/typestate.py]: Maybe rename these to `subtype_caches` and `proper_subtype_caches` (then the above comment becomes moot). Maybe also use `_` prefix since these should only be accessed through methods.

File: mypy/typestate.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    caches_proper = {}  # type: ClassVar[Dict[TypeInfo, Set[Tuple[Instance, Instance]]]]
+
+    @classmethod
+    def reset_all(cls) -> None:
```
Comment (JukkaL) [inline mypy/typestate.py]: Maybe rename to `clear` or `clear_caches`?

File: mypy/typestate.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        cls.caches = {}
+        cls.caches_proper = {}
+
+    @classmethod
+    def add_caches(cls, info: 'TypeInfo') -> None:
```
Comment (JukkaL) [inline mypy/typestate.py]: As discussed above, this seems unnecessary in the public API. We can create the caches on demand here.

File: mypy/typestate.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        cls.caches_proper[info] = set()
+
+    @classmethod
+    def reset_caches(cls, info: 'TypeInfo') -> None:
+        cls.caches[info].clear()
+        cls.caches_proper[info].clear()
+
+    @classmethod
+    def is_cached_subtype_check(cls, left: 'Instance', right: 'Instance',
+                                proper_subtype: bool = False) -> bool:
```
Comment (JukkaL) [inline mypy/typestate.py]: Style nit: I think that it would be better to have two methods, one for subtype checks and the other for proper subtype checks. Bool mode arguments are a minor code smell in Python.

File: mypy/typestate.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        info = right.type
+        if proper_subtype:
+            cache = cls.caches_proper[info]
+        else:
+            cache = cls.caches[info]
+        return (left, right) in cache
+
+    @classmethod
+    def record_subtype_cache_entry(cls, left: 'Instance', right: 'Instance',
+                                   proper_subtype: bool = False) -> None:
```
Comment (JukkaL) [inline mypy/typestate.py]: Like above, I'd prefer two variants of this method.

File: mypy/dmypy_server.py [hunk 13]
```diff
@@ -152,6 +153,7 @@ def serve(self) -> None:
                         conn, addr = sock.accept()
                     except socket.timeout:
                         print("Exiting due to inactivity.")
+                        TypeState.reset_all_subtype_caches()
```
Comment (JukkaL) [inline mypy/dmypy_server.py]: Maybe move this call to a method (say, `free_global_state`) since this call and the identical one below must stay in sync, and by having them in a single method makes that easier.

File: mypy/dmypy_server.py [hunk 14]
```diff
@@ -182,6 +184,7 @@ def serve(self) -> None:
                     conn.close()
                     if command == 'stop':
                         sock.close()
+                        TypeState.reset_all_subtype_caches()
```
Comment (JukkaL) [inline mypy/dmypy_server.py]: This is the other call I mentioned above.

File: mypy/nodes.py [hunk 15]
```diff
@@ -2133,7 +2128,7 @@ def calculate_mro(self) -> None:
         self.mro = mro
         # The property of falling back to Any is inherited.
         self.fallback_to_any = any(baseinfo.fallback_to_any for baseinfo in self.mro)
-        self.reset_subtype_cache()
+        TypeState.reset_all_subtype_caches_for(self)
```
Comment (JukkaL) [inline mypy/nodes.py]: Bummer that we still have the circular dependency. We could move the `calculate_mro` method to another module (that `mypy.node` doesnt't depend on) to get rid of it. Can you do it or at least add a TODO comment about this refactor?

File: mypy/build.py [hunk 16]
```diff
@@ -265,9 +265,11 @@ def _build(sources: List[BuildSource],
                            flush_errors=flush_errors,
                            fscache=fscache)
 
+    TypeState.reset_all_subtype_caches()
     try:
-        TypeState.reset_all()
         graph = dispatch(sources, manager)
+        if not options.cache_fine_grained or options.fine_grained_incremental:
```
Comment (JukkaL) [inline mypy/build.py]: Should we actually clear the caches when using `cache_fine_grained`? I think that it only generates some additional module metadata that is not related to subtype checks.

Comment (ilevkivskyi) [reply mypy/build.py]: Indeed, we don't need it.
